datafusion/proto/src/physical_plan/to_proto.rs

```diff
@@ -196,6 +216,209 @@ impl TryFrom<Arc<dyn AggregateExpr>> for protobuf::PhysicalExprNode {
+                }
+                if window_shift_expr.get_shift_offset() >= 0 {
+                    protobuf::BuiltInWindowFunction::Lag
+                } else {
+                    protobuf::BuiltInWindowFunction::Lead
+                }
+            } else if let Some(nth_value_expr) =
+                built_in_fn_expr.downcast_ref::<NthValue>()
+            {
+                match nth_value_expr.get_kind() {
+                    NthValueKind::First => protobuf::BuiltInWindowFunction::FirstValue,
+                    NthValueKind::Last => protobuf::BuiltInWindowFunction::LastValue,
+                    NthValueKind::Nth(n) => {
+                        args.insert(
+                            1,
+                            Arc::new(Literal::new(
+                                datafusion_common::ScalarValue::Int64(Some(n as i64)),
+                            )),
+                        );
+                        protobuf::BuiltInWindowFunction::NthValue
+                    }
+                }
+            } else {
+                return not_impl_err!("BuiltIn function not supported: {expr:?}");
+            };
+
+            physical_window_expr_node::WindowFunction::BuiltInFunction(builtin_fn as i32)
+        } else if let Some(plain_aggr_window_expr) =
+            expr.downcast_ref::<PlainAggregateWindowExpr>()
+        {
+            aggr_expr_to_window_fn(plain_aggr_window_expr.get_aggregate_expr().as_ref())?
+        } else if let Some(sliding_aggr_window_expr) =
+            expr.downcast_ref::<SlidingAggregateWindowExpr>()
+        {
+            aggr_expr_to_window_fn(
+                sliding_aggr_window_expr.get_aggregate_expr().as_ref(),
+            )?
+        } else {
+            return not_impl_err!("WindowExpr not supported: {window_expr:?}");
+        };
+
+        let args = args
+            .into_iter()
+            .map(|e| e.try_into())
+            .collect::<Result<Vec<protobuf::PhysicalExprNode>>>()?;
+
+        let partition_by = window_expr
+            .partition_by()
+            .iter()
+            .map(|p| p.clone().try_into())
+            .collect::<Result<Vec<protobuf::PhysicalExprNode>>>()?;
+
+        let order_by = window_expr
+            .order_by()
+            .iter()
+            .map(|o| o.clone().try_into())
+            .collect::<Result<Vec<protobuf::PhysicalSortExprNode>>>()?;
+
+        let window_frame: protobuf::WindowFrame = window_expr
+            .get_window_frame()
+            .as_ref()
+            .try_into()
+            .map_err(|e| DataFusionError::Internal(format!("{e}")))?;
+
+        let name = window_expr.name().to_string();
+
+        Ok(protobuf::PhysicalWindowExprNode {
+            args,
+            partition_by,
+            order_by,
+            window_frame: Some(window_frame),
+            window_function: Some(window_function),
+            name,
+        })
+    }
+}
+
+fn aggr_expr_to_window_fn(
```

Comment:
Something doesn't seem right here -- it seems like this list replicates the list in https://github.com/apache/arrow-datafusion/blob/c26373ab5c5385fa8da9757c36befadb57bce3e2/datafusion/proto/src/physical_plan/to_proto.rs#L98-L121

Maybe we could refactor the conversion to avoid the duplication

Reply:
Yeah, i saw that. Will refactor and combine both the funcs.

Reply:
Thank you@